src/cirrus/lib2/process_payload.py

```diff
@@ -0,0 +1,384 @@
+            raise ValueError('ProcessPayload.process must have `upload_options` defined')
+
+        if 'workflow' not in self.process:
+            raise ValueError(
+                'ProcessPayload.process must have `workflow` specifying workflow name'
+            )
+
+        # convert old functions field to tasks
+        if 'functions' in self.process:
+            warnings.warn("Deprecated: process 'functions' has been renamed to 'tasks'")
+            self.process['tasks'] = self.process.pop('functions')
+
+        if 'tasks' not in self.process:
+            raise ValueError('ProcessPayload.process must have `tasks` defined')
+
+        self.tasks = self.process['tasks']
+
+        if 'workflow-' not in self['id']:
+            raise ValueError(f'Invalid payload id: {self["id"]}')
+
+        # TODO - validate with a JSON schema
+        #if schema:
+        #    pass
+        # For now, just make check that there is at least one item
+        if len(self['features']) < 1:
+            raise ValueError(
+                'ProcessPayload.features is empty; must have at least one feature'
+            )
+
+        for item in self['features']:
+            if 'links' not in item:
+                item['links'] = []
+
+        # update collection IDs of member Items
+        self.assign_collections()
+
+        self.state_item = state_item
+
+    @classmethod
+    def from_event(cls, event: Dict, **kwargs) -> ProcessPayload:
+        """Parse a Cirrus event and return a ProcessPayload instance
+
+        Args:
+            event (Dict): An event from SNS, SQS, or containing an s3 URL to payload
+
+        Returns:
+            ProcessPayload: A ProcessPayload instance
+        """
+        records = list(extract_event_records(event))
+
+        if len(records) == 0:
+            raise ValueError('Failed to extract record: %s', json.dumps(event))
+        elif len(records) > 1:
+            raise ValueError('Multiple payloads are not supported')
+
+        return cls(records[0], **kwargs)
+
+    def get_task(self, task_name, *args, **kwargs):
+        return self.tasks.get(task_name, *args, **kwargs)
+
+    def next_payloads(self):
+        if isinstance(self['process'], dict) or len(self['process']) <= 1:
+            return None
+        next_processes = (
+            [self['process'][1]]
+            if isinstance(self['process'][1], dict)
+            else self['process'][1]
+        )
+        for process in next_processes:
+            new = deepcopy(self)
+            del new['id']
+            new['process'].pop(0)
+            new['process'][0] = process
+            if 'chain_filter' in process:
+                jsonfilter = jsonpath.parse(
+                    f'$.features[?({process["chain_filter"]})]',
+                )
+                new['features'] = [
+                    match.value for match in jsonfilter.find(new)
+                ]
+            yield new
+
+    def update(self):
```

Comment:
This should be `create_id`? Should probably exit before doing all this stuff if `id` set.

This should raise an error if no `id` and no `features` or something else goes wrong making the `id`.